File: articles/flow-and-optimal-experience-positive-psychology/fortran/flow_indices.f90
program flow_indices
  implicit none
  real :: challenge_level, skill_level, attention_focus, feedback_quality, goal_clarity
  real :: task_meaning, autonomy_support, distraction_load, interruption_count
  real :: flow_score, performance_score, learning_gain, fatigue_score, recovery_quality, wellbeing_score
  real :: balance, attention_ecology, engagement_context, sustainable_flow

  challenge_level = 7.1
  skill_level = 7.0
  attention_focus = 7.3
  feedback_quality = 7.1
  goal_clarity = 7.4
  task_meaning = 7.4
  autonomy_support = 7.2
  distraction_load = 3.0
  interruption_count = 1.0
  flow_score = 7.2
  performance_score = 7.3
  learning_gain = 0.39
  fatigue_score = 3.2
  recovery_quality = 7.0
  wellbeing_score = 7.2

  balance = -abs(challenge_level - skill_level)
  attention_ecology = attention_focus + feedback_quality + goal_clarity - distraction_load - interruption_count
  engagement_context = balance + attention_focus + feedback_quality + goal_clarity + &
      task_meaning + autonomy_support - distraction_load - interruption_count
  sustainable_flow = flow_score + task_meaning + autonomy_support + recovery_quality - &
      fatigue_score - distraction_load

  print *, "Challenge-skill balance:", balance
  print *, "Attentional ecology:", attention_ecology
  print *, "Deep engagement context:", engagement_context
  print *, "Sustainable flow index:", sustainable_flow
end program flow_indices

File: articles/flow-and-optimal-experience-positive-psychology/c/flow_indices.c
#include <stdio.h>
#include <math.h>

typedef struct {
    double challenge_level;
    double skill_level;
    double attention_focus;
    double feedback_quality;
    double goal_clarity;
    double task_meaning;
    double autonomy_support;
    double distraction_load;
    double interruption_count;
    double flow_score;
    double performance_score;
    double learning_gain;
    double fatigue_score;
    double recovery_quality;
    double wellbeing_score;
} FlowIndicators;

double balance_index(FlowIndicators x) {
    return -fabs(x.challenge_level - x.skill_level);
}

double attentional_ecology(FlowIndicators x) {
    return x.attention_focus + x.feedback_quality + x.goal_clarity - x.distraction_load - x.interruption_count;
}

double deep_engagement_context(FlowIndicators x) {
    return balance_index(x) + x.attention_focus + x.feedback_quality + x.goal_clarity +
           x.task_meaning + x.autonomy_support - x.distraction_load - x.interruption_count;
}

double sustainable_flow_index(FlowIndicators x) {
    return x.flow_score + x.task_meaning + x.autonomy_support + x.recovery_quality -
           x.fatigue_score - x.distraction_load;
}

int main(void) {
    FlowIndicators example = {7.1, 7.0, 7.3, 7.1, 7.4, 7.4, 7.2, 3.0, 1.0, 7.2, 7.3, 0.39, 3.2, 7.0, 7.2};

    printf("Challenge-skill balance: %.3f\n", balance_index(example));
    printf("Attentional ecology: %.3f\n", attentional_ecology(example));
    printf("Deep engagement context: %.3f\n", deep_engagement_context(example));
    printf("Sustainable flow index: %.3f\n", sustainable_flow_index(example));

    return 0;
}
